### PR TITLE
Rewrite stack2nix

### LINKED ISSUE #37: Expose output from cabal2nix calls for dependencies with subdirs
To support the `subdirs` directive in stack.yaml for non-local dependencies [this hack](https://github.com/input-output-hk/stack2nix/blob/master/src/Stack2nix.hs#L223-L224) was introduced. When there's a single dependency with multiple `subdirs` some output and retry logic gets suppressed.

### PR DESCRIPTION
**Pull request by domenkozar**
- get rid of hnix and use Derivation type to achieve test/haddock filtering #81 
- use source stackage plan instead of build plan #70 
- allow overriding base GHC package #51 

This allows me to build stack2nix and cardano-sl without any issues.
  
 Also fixes #37 and #74

### CODE REVIEW

**Comment by kirelagin** [comment]
I’m not sure that this actually fixes #74 🤔. I’m on macOS, so for me the issue is manifested in a slightly different way.

I basically did the simplest thing:

```
nix-shell --pure -p nix git cabal-install cabal2nix nix-prefetch-git darwin.security_tool \
  --run "NIX_PATH=$NIX_PATH ~/.local/bin/stack2nix ."
```

(Here `~/.local/bin/stack2nix` is `stack2nix` built and installed with `stack install` at 9ff935c94d0f7efcb961926f745f9ca85c7f378b.)

What happens is `stack` proceeds to build GHC and for me it fails as it can’t find `perl` (the shell is pure), but on NixOS tripping on `/bin/bash` is no surprise.

On the other hand, when I run:

```
nix-shell --pure -p nix stack --run "NIX_PATH=$NIX_PATH stack build"
```

it doesn’t try to build GHC, I suspect because it takes it from nix.

**Comment by domenkozar** [comment]
@kirelagin what project are you running this on?

**Comment by kirelagin** [comment]
Well, it’s a private one. Let me try to come up with a minimal example.

**Comment by kirelagin** [comment]
It happens even for a project consisting of two files:

stack.yaml:

```yaml
resolver: lts-9.3
packages:
  - .
```

package.yaml:

```yaml
name: test
```

**Comment by kirelagin** [comment]
Actually, just `stack.yaml` is enough. `stack` tries to build GHC first thing, before even trying to read the package.

**Comment by kirelagin** [comment]
In case it helps, the output is:

```
Ensuring git version is >= 2 ...
Ensuring cabal version is >= 2 ...
Preparing to install GHC to an isolated location.
This will not interfere with any system-level installation.
Preparing to download ghc-8.0.2 ...
Already downloaded.
Unpacking GHC into /private/tmp/s2n/programs/x86_64-osx/ghc-8.0.2.temp/ ...
^C
```

**Comment by kirelagin** [comment]
Meanwhile, another issue with this PR is that `stack2nix` now ignores `-o`.

**Comment by domenkozar** [comment]
@kirelagin you're right, it still doesn't use Nix with stack properly.

**Comment by domenkozar** [comment]
Thanks @kirelagin, I've fixed `-o`. I understand what's going on with ghc, but need to think further how to properly solve it.

**Comment by domenkozar** [comment]
CI fails due to #74 now, I have a workaround in mind.

**Comment by domenkozar** [comment]
@kirelagin I've pushed a fix, can you try now with: `$ nix-shell --pure -p nix glibcLocales --run "LANG=en_US.utf-8 NIX_REMOTE=daemon NIX_PATH=$NIX_PATH ~/.local/bin/stack2nix ."`

File: src/Stack2nix.hs [hunk 1]
```diff
@@ -26,7 +26,13 @@ import           System.IO.Temp             (withSystemTempDirectory)
 
 stack2nix :: Args -> IO ()
 stack2nix args@Args{..} = do
-  checkRuntimeDeps
+  ensureExecutableExists "cabal" "cabal-install"
+  ensureExecutableExists "git" "git"
+  ensureExecutableExists "nix-prefetch-git" "nix-prefetch-scripts"
+  ensureExecutableExists "ghc" "haskell.compiler.ghc802"
```
Comment (kirelagin) [inline src/Stack2nix.hs]: 🤔 
Weeeeell, I mean... Yeah, this will work, but why exactly 8.0.2? 😒

File: src/Stack2nix/Util.hs [hunk 2]
```diff
@@ -51,3 +56,18 @@ logDebug :: Args -> String -> IO ()
 logDebug args msg
   | argVerbose args = hPutStrLn stderr msg
   | otherwise = return ()
+
+
+ensureExecutableExists :: String -> String -> IO ()
+ensureExecutableExists executable nixAttr = do
```
Comment (kirelagin) [inline src/Stack2nix/Util.hs]: So, you basically reimplemented `nix-shell` :).

**Comment by domenkozar** [comment]
@kirelagin see https://github.com/input-output-hk/stack2nix/issues/84 - 8.0.x is assumed at other places as well.

**Comment by kirelagin** [comment]
Ok, got it.
Regarding your fix, it helps only when the compiler is not present at all, but not if the version is wrong:

```
stack2nix: Compiler version mismatched, found ghc-8.2.2 (x86_64), but expected minor version match with ghc-8.0.2 (x86_64)
```

**Pull request merged**